mongo/integration/collection_test.go

```diff
@@ -1067,6 +1067,68 @@ func TestCollection(t *testing.T) {
+			}
+			_, err := capped.BulkWrite(mtest.Background, models, options.BulkWrite())
+			assert.Nil(t, err, "BulkWrite error: %v", err)
+
+			models = []mongo.WriteModel{
+				mongo.NewDeleteOneModel().SetFilter(bson.D{{"_id", "id0"}}),
+				mongo.NewDeleteManyModel().SetFilter(bson.D{{"_id", "id0"}}),
+				mongo.NewUpdateManyModel().SetFilter(bson.D{{"_id", "id3"}}).SetUpdate(bson.D{{"$set", bson.D{{"_id", 3.14159}}}}),
+				mongo.NewInsertOneModel().SetDocument(bson.D{{"_id", "id1"}}),
+				mongo.NewUpdateOneModel().SetFilter(bson.D{{"_id", "id3"}}).SetUpdate(bson.D{{"$set", bson.D{{"_id", 3.14159}}}}),
+				mongo.NewReplaceOneModel().SetFilter(bson.D{{"_id", "id3"}}).SetReplacement(bson.D{{"_id", 3.14159}}),
+			}
+			_, err = capped.BulkWrite(mtest.Background, models, options.BulkWrite().SetOrdered(false))
+			bwException, ok := err.(mongo.BulkWriteException)
+			assert.True(mt, ok, "expected error of type %T, got %T", mongo.BulkWriteException{}, err)
+
+			assert.Equal(mt, len(bwException.WriteErrors), 6, "expected 6 writeErrors, got %v", len(bwException.WriteErrors))
+			for _, writeErr := range bwException.WriteErrors {
+				switch writeErr.Request.(type) {
+				case *mongo.DeleteOneModel:
+					assert.Equal(mt, writeErr.Index, 0, "expected index 0, got %v", writeErr.Index)
+				case *mongo.DeleteManyModel:
+					assert.Equal(mt, writeErr.Index, 1, "expected index 1, got %v", writeErr.Index)
+				case *mongo.UpdateManyModel:
+					assert.Equal(mt, writeErr.Index, 2, "expected index 2, got %v", writeErr.Index)
+				case *mongo.InsertOneModel:
+					assert.Equal(mt, writeErr.Index, 3, "expected index 3, got %v", writeErr.Index)
+				case *mongo.UpdateOneModel:
+					assert.Equal(mt, writeErr.Index, 4, "expected index 4, got %v", writeErr.Index)
+				case *mongo.ReplaceOneModel:
+					assert.Equal(mt, writeErr.Index, 5, "expected index 5, got %v", writeErr.Index)
+				}
+
+			}
+		})
+		mt.Run("unordered upsertID index", func(mt *mtest.T) {
+			id1 := "id1"
+			id3 := "id3"
+			models := []mongo.WriteModel{
+				mongo.NewDeleteOneModel().SetFilter(bson.D{{"_id", "id0"}}),
+				mongo.NewReplaceOneModel().SetFilter(bson.D{{"_id", id1}}).SetReplacement(bson.D{{"_id", id1}}).SetUpsert(true),
+				mongo.NewDeleteOneModel().SetFilter(bson.D{{"_id", "id2"}}),
+				mongo.NewReplaceOneModel().SetFilter(bson.D{{"_id", id3}}).SetReplacement(bson.D{{"_id", id3}}).SetUpsert(true),
+				mongo.NewDeleteOneModel().SetFilter(bson.D{{"_id", "id4"}}),
+			}
+			res, err := mt.Coll.BulkWrite(mtest.Background, models, options.BulkWrite().SetOrdered(false))
+			assert.Nil(mt, err, "bulkwrite error: %v", err)
+
+			assert.Equal(mt, len(res.UpsertedIDs), 2, "expected 2 UpsertedIDs, got %v", len(res.UpsertedIDs))
+			assert.Equal(mt, res.UpsertedIDs[1].(string), id1, "expected UpsertedIDs[1] to be %v, got %v", id1, res.UpsertedIDs[1])
+			assert.Equal(mt, res.UpsertedIDs[3].(string), id3, "expected UpsertedIDs[3] to be %v, got %v", id3, res.UpsertedIDs[3])
+		})
```

Comment:
I was also going to ask whether there are tests of individual commands being split into multiple batches. Though that was addressed in GODRIVER-1363 and appears to be tested in https://github.com/mongodb/mongo-go-driver/commit/c814cfb676dc29a6ad6171e138510ee88c51983f#diff-1572fa8947f9463542abdc5cb3a11528R204.

Reply:
that will be handled in GODRIVER-1715